client-runtime/src/main/java/com/microsoft/rest/v2/RestProxy.java

```diff
@@ -84,19 +94,61 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
             }
         }
 
-        final HttpResponse response = httpClient.sendRequest(request);
-
-        final Class<?> returnType = method.getReturnType();
-        if (returnType.equals(Void.TYPE) || !response.hasBody()) {
-            return null;
-        } else if (returnType.isAssignableFrom(InputStream.class)) {
-            return response.bodyAsInputStream();
-        } else if (returnType.isAssignableFrom(byte[].class)) {
-            return response.bodyAsByteArray();
-        } else {
-            final String responseBodyString = response.bodyAsString();
-            return serializer.deserialize(responseBodyString, returnType);
+        Object result = null;
+        if (!methodDetails.isAsync()) {
+            final HttpResponse response = httpClient.sendRequest(request);
+
+            final Class<?> returnType = methodDetails.getReturnType();
+            if (returnType.equals(Void.TYPE) || !response.hasBody() || methodDetails.method().equalsIgnoreCase("HEAD")) {
+                result = null;
+            } else if (returnType.isAssignableFrom(InputStream.class)) {
+                result = response.bodyAsInputStream();
+            } else if (returnType.isAssignableFrom(byte[].class)) {
+                result = response.bodyAsByteArray();
+            } else {
+                final String responseBodyString = response.bodyAsString();
+                result = serializer.deserialize(responseBodyString, returnType);
+            }
+        }
+        else {
+            final Single<? extends HttpResponse> asyncResponse = httpClient.sendRequestAsync(request);
+            final Class<?> methodReturnType = method.getReturnType();
+            if (methodReturnType.equals(Single.class)) {
+                result = asyncResponse.flatMap(new Func1<HttpResponse, Single<?>>() {
+                    @Override
+                    public Single<?> call(HttpResponse response) {
+                        Single<?> asyncResult;
+                        final Class<?> singleReturnType = methodDetails.getReturnType();
+                        if (methodDetails.method().equalsIgnoreCase("HEAD")) {
+                            asyncResult = Single.just(null);
```

Comment:
In Azure, if the response is boolean, we return true for 2xx responses and false otherwise. You can also do this in the Azure flavored parser.